torchtext/datasets/unsupervised_learning.py

```diff
@@ -0,0 +1,138 @@
+             (r'J', 'j'), (r'K', 'k'), (r'L', 'l'),
+             (r'M', 'm'), (r'N', 'n'), (r'O', 'o'),
+             (r'P', 'p'), (r'Q', 'q'), (r'R', 'r'),
+             (r'S', 's'), (r'T', 't'), (r'U', 'u'),
+             (r'V', 'v'), (r'W', 'w'), (r'X', 'x'),
+             (r'Y', 'y'), (r'Z', 'z'),
+             (r'0', ' zero '), (r'1', ' one '), (r'2', ' two '),
+             (r'3', ' three '), (r'4', ' four '), (r'5', ' five '),
+             (r'6', ' six '), (r'7', ' seven '), (r'8', ' eight '),
+             (r'9', ' nine '),
+             (r'[^a-z\n]+', ' '),
+             (r'\n ', ''),
+             (r'\s+', ' '),
+             (r'\n\s*\n', r'\n')
+             ]
+enwik9_norm_transform = custom_replace(_patterns)
+
+
+def generate_offsets(filename):
+    offsets = []
+    with open(filename) as f:
+        offsets.append(f.tell())
+        while f.readline():
+            offsets.append(f.tell())
+    return offsets
+
+
+def read_lines_from_iterator(data_path, offsets, begin_line, num_lines):
+    with open(data_path) as f:
+        f.seek(offsets[begin_line])
+        for i in range(num_lines):
+            yield f.readline()
+
+
+def normalized_raw_enwik9(input_filename, output_filename):
+    with open(input_filename, 'r') as f1:
+        with open(output_filename, 'w') as f2:
+            while True:
+                line = f1.readline()
+                if not line:
+                    break
+                line = list(enwik9_norm_transform([line]))[0]
+                if line != ' ' and line != '':
+                    if line[0] == ' ':
+                        line = line[1:]
+                    f2.writelines(line + '\n')
+    return
+
+
+class EnWik9(torch.utils.data.Dataset):
+    r"""Compressed size of first 10^9 bytes of enwiki-20060303-pages-articles.xml.
+        It's part of Large Text Compression Benchmark project
+    """
+
+    def __init__(self, begin_line=0, num_lines=6348957, root='.data'):
+        """Initiate EnWik9 dataset.
+
+        Arguments:
+            begin_line: the number of beginning line. Default: 0
+            num_lines: the number of lines to be loaded. Default: 6348957
+            root: Directory where the datasets are saved. Default: ".data"
+            data: a list of label/tokens tuple. tokens are a tensor after
+
+        Examples:
+            >>> from torchtext.datasets import EnWik9
+            >>> enwik9 = EnWik9(num_lines=20000)
+            >>> vocab = enwik9.get_vocab()
+
+        """
+
+        super(EnWik9, self).__init__()
+
+        processed_file = os.path.join(root, 'norm_enwik9')
+        if not os.path.exists(processed_file):
+            url = 'http://mattmahoney.net/dc/enwik9.zip'
+            dataset_zip = download_from_url(url,
+                                            path=os.path.join(root, 'enwik9.zip'),
+                                            root=root)
+            extracted_file = extract_archive(dataset_zip)
```

Comment:
if the data is already downloaded or extracted, could we want to skip these steps?

Reply:
the download function will automatically skip if the downloaded file is detected.

Reply:
If the file was downloaded then extracted, and then the user deletes the downloaded file, it will be downloaded again even though it's already extracted, right?

Reply:
Yes. If downloaded file is deleted, then it will be downloaded again. But I have a `if` check above to see if the processed_file exists. And in this case, downloading is much more expensive than extracting.